src/main/java/org/sonatype/nexus/repository/r/internal/RPackagesBuilder.java

```diff
@@ -0,0 +1,118 @@
+ */
+package org.sonatype.nexus.repository.r.internal;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeMap;
+
+import org.sonatype.nexus.repository.storage.Asset;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.util.Collections.unmodifiableMap;
+import static org.sonatype.nexus.repository.r.internal.RAttributes.P_DEPENDS;
+import static org.sonatype.nexus.repository.r.internal.RAttributes.P_IMPORTS;
+import static org.sonatype.nexus.repository.r.internal.RAttributes.P_LICENSE;
+import static org.sonatype.nexus.repository.r.internal.RAttributes.P_NEEDS_COMPILATION;
+import static org.sonatype.nexus.repository.r.internal.RAttributes.P_PACKAGE;
+import static org.sonatype.nexus.repository.r.internal.RAttributes.P_SUGGESTS;
+import static org.sonatype.nexus.repository.r.internal.RAttributes.P_VERSION;
+
+/**
+ * Builds the contents of a PACKAGES file based on the provided assets, taking into account the greatest version of a
+ * particular package that is available in a (hosted) repository.
+ *
+ * Note that this maintains all pertinent information for the "latest" version of each package in memory, though the
+ * actual amount of information for each package is rather small.
+ *
+ * TODO: Ensure that we build this information and actually cache it, rather than doing this for each request.
+ */
+public class RPackagesBuilder
+{
+  /**
+   * The greatest version of each package currently encountered during this run.
+   */
+  private final Map<String, RPackageVersion> packageVersions = new HashMap<>();
+
+  /**
+   * The package information as of the last time we processed the greatest version of this package. A {@code TreeMap} is
+   * used to maintain ordering by package name.
+   */
+  private final Map<String, Map<String, String>> packageInformation = new TreeMap<>();
+
+  /**
+   * The path to the PACKAGES file that is being generated.
+   */
+  private final String packagesPath;
+
+  /**
+   * Constructor.
+   *
+   * @param packagesPath The path to the PACKAGES file that is being generated.
+   */
+  public RPackagesBuilder(final String packagesPath) {
+    this.packagesPath = checkNotNull(packagesPath);
+  }
+
+  /**
+   * Processes an asset, updating the greatest version and details for the package if appropriate.
+   *
+   * @param asset The asset to process.
+   */
+  public void append(final Asset asset) {
+    // is this asset at this particular path?
+    String base = basePath(packagesPath);
+    if (base.equals(basePath(asset.name()))) {
+
+      // is this a newer (or equal) version of this asset's package than the one we currently have?
+      String packageName = asset.formatAttributes().get(P_PACKAGE, String.class);
+      RPackageVersion newVersion = new RPackageVersion(asset.formatAttributes().get(P_VERSION, String.class));
+      RPackageVersion oldVersion = packageVersions.getOrDefault(packageName, newVersion);
+      if (newVersion.compareTo(oldVersion) >= 0) {
```

Comment:
There is probably no need to check for >=, just > 0 should suffice

Reply:
There's the scenario where you don't have an initial occurrence of _any_ version for that package, so this was my way of getting around it _sans_ a separate existence or null check.

(That said, if you didn't notice that in reviewing it, it may also be that this is too cryptic a means of implementing that, so maybe we do want an explicit check. I kind of like this in that it deals with that scenario and also keeps down the total number of conditionals.)

Reply:
How do you feel about https://github.com/sonatype/nexus-repository-r/pull/18/commits/e0a43d9b8932856a24c07e70db0f259e2841c735 as an alternative? (Reading it now, I think this is probably the clearer one.)

Reply:
Much clearer and thanks for the original explanation